optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,330 @@
+import sys
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Sequence
+
+import numpy as np
+
+import optuna
+from optuna import logging
+from optuna._experimental import experimental
+from optuna._imports import _LazyImport
+from optuna._transform import _SearchSpaceTransform
+from optuna.distributions import BaseDistribution
+from optuna.distributions import CategoricalDistribution
+from optuna.samplers import BaseSampler
+from optuna.study import Study
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+_logger = logging.get_logger(__name__)
+
+_SUGGESTED_STATES = (TrialState.COMPLETE, TrialState.PRUNED)
+
+
+@experimental("3.0.0")
+class QMCSampler(BaseSampler):
+    """A Quasi Monte Carlo Sampler that generates low-discrepancy sequences.
+
+    Quasi Monte Carlo (QMC) sequences are designed to have lower discrepancies than
+    standard random seqeunces. They are known to perform better than the standard
+    randam sequences in hyperparameter optimization.
+
+    For further information about the use of QMC sequences for hyperparameter optimization,
+    please refer to the following paper:
+
+    - `Bergstra, James, and Yoshua Bengio. Random search for hyper-parameter optimization.
+      Journal of machine learning research 13.2, 2012.
+      <https://jmlr.org/papers/v13/bergstra12a.html>`_
+
+    We use the QMC implementations in Scipy. For the details of the QMC algorithm,
+    see the Scipy API references on `scipy.stats.qmc
+    <https://scipy.github.io/devdocs/reference/stats.qmc.html>`_.
+
+    .. note:
+        If your search space contains categorical parameters, it samples the catagorical
+        parameters by its `independent_sampler` without using QMC algorithm.
```

Comment:
What is the limitation here? You can use modulo on a given dimension to select a given category. It would be better than using a normal random sampler.

Reply:
We have discretized uniform/log-uniform distributions and this can be a replacement of categorical distributions when we have categorical variables which have similarities between categories (e.g. ordered categorical variables).

Reply:
Ok but that's my point, you could still use QMC for that. Otherwise you weaken the sample.

Reply:
FYI, I took a benchmark and could not observe any difference in performance between QMC based categorical and random categorical sampling. The benchmark results and some discussion is at https://github.com/optuna/optuna/pull/2423#discussion_r588111091 around March 3-5, 2021.

Reply:
OK, I think I got your point. When we use random samples for categorical samples, we can have higher-order performance only for cases where the effect of categorical variables on the objective value is quite regular (e.g. additive). When we use some dimensions of QMC to sample the categorical variables, we can have higher-order convergence for a more general function class I guess. It might be better to sample categorical variables from QMC even if they do not improve the benchmark performance.